cpp/src/arrow/engine/substrait/expression_internal.cc

```diff
@@ -138,6 +138,77 @@ std::string EnumToString(int value, const google::protobuf::EnumDescriptor* desc
   return value_desc->name();
 }
 
+Result<compute::Expression> FromProto(const substrait::Expression::ReferenceSegment* ref,
+                                      const ExtensionSet& ext_set,
+                                      const ConversionOptions& conversion_options,
+                                      std::optional<compute::Expression> in_expr) {
+  auto in_ref = ref;
+  auto& out = in_expr;
+  while (ref != nullptr) {
+    switch (ref->reference_type_case()) {
+      case substrait::Expression::ReferenceSegment::kStructField: {
+        auto index = ref->struct_field().field();
+        if (!out) {
+          // Root StructField (column selection)
+          out = compute::field_ref(FieldRef(index));
+        } else if (auto out_ref = out->field_ref()) {
+          // Nested StructFields on the root (selection of struct-typed column
+          // combined with selecting struct fields)
+          out = compute::field_ref(FieldRef(*out_ref, index));
+        } else if (out->call() && out->call()->function_name == "struct_field") {
+          // Nested StructFields on top of an arbitrary expression
+          auto* field_options =
+              checked_cast<compute::StructFieldOptions*>(out->call()->options.get());
+          field_options->field_ref = FieldRef(std::move(field_options->field_ref), index);
+        } else {
+          // First StructField on top of an arbitrary expression
+          out = compute::call("struct_field", {std::move(*out)},
+                              arrow::compute::StructFieldOptions({index}));
+        }
```

Comment:
The code here is perhaps more complex than the caller needs in many cases.  Ideally it would return a simple field reference.  The challenge is that Arrow's expressions don't support field references into intermediate outputs.  For example, if you have a function `min_max` that returns a struct like `{ "minimum": 0, "maximum": 10} }` then there is no way to say `min_max(x).minimum > 0` using a field reference.  So you end up needing to use the `struct_field` function to create `call("gt", { call("struct_field", { field_ref("x"), "minimum" }), 0 })`.

In many cases (e.g. aggregate keys, join keys, sort keys) Acero will only accept "direct references".  These will always end up being expressible as `FieldRef`.

So, there may be some value in creating a second `DirectReferenceFromProto` that accepts `ReferenceSegment` and returns `FieldRef`.  The implementation could call the existing `FromProto(Expression)`:

```
Result<FieldRef> DirectReferenceFromProto(const substrait::Expression& substrait_expr,
                                          const ExtensionSet& ext_set,
                                          const ConversionOptions& conversion_options) {
  ARROW_ASSIGN_OR_RAISE(compute::Expression expr,
                        FromProto(substrait_expr, ext_set, conversion_options));
  const FieldRef* field_ref = expr.field_ref();
  if (field_ref) {
    return *field_ref;
  } else {
    return Status::Invalid(
        "A direct reference was expected but a more complex expression was given "
        "instead");
  }
}
```

You would still need the change here so I agree with breaking it out into its own method.